### PR TITLE
MAINT: Consistent method keywords

### PR DESCRIPTION
**Pull request by kygoben**
From issue #1187 used the descriptions from @mtd91429 to make a few modifications to PdfWriter and PdfReader

* `indirect_reference` ➔ `indirect_ref` and `ido`  ➔ `indirect_ref` in `PdfReader._get_object_from_stream` and `PdfWriter.get_object`
* `dest`➔ `pagedest` and  `dest_ref` ➔ `pagedest_ref` in `PdfWriter.add_outline_item_destination` and `PdfWriter.add_named_destination_object`

### CODE REVIEW

**Comment by kygoben** [comment]
Hi @MartinThoma, I am a bit unsure how to completely approach your comments. For instance, you said that the migration guide would need to be updated. Where can I find this, and how should this be updated? And the second one is to give people a warning when using the keywords. Thank you for your help :)

**Comment by kygoben** [comment]
Not sure what I did incorrectly in these last modifications. Could someone give me some pointers to fix the test outcomes? 
Thanks,
Kyle

**Comment by MartinThoma** [comment]
I'll have a look at the weekend. I was at a company event this week and moving/getting married last week :-)

**Comment by MartinThoma** [comment]
I'll add this either before the PyPDF2==3.0.0 release or earlier.

File: PyPDF2/_writer.py [hunk 1]
```diff
@@ -201,10 +201,10 @@ def _add_object(self, obj: PdfObject) -> IndirectObject:
         self._objects.append(obj)
         return IndirectObject(len(self._objects), 0, self)
 
-    def get_object(self, ido: IndirectObject) -> PdfObject:
-        if ido.pdf != self:
+    def get_object(self, indirect_reference: IndirectObject) -> PdfObject:
```
Comment (MartinThoma) [inline PyPDF2/_writer.py:204]:
```suggestion
    def get_object(self, indirect_reference: Optional[IndirectObject]=None, ido: Optional[IndirectObject]=None) -> PdfObject:
        if ido is not None:
            if indirect_reference is not None:
                raise ValueError("Please only set 'indirect_reference'. The 'ido' argument is deprecated.")
            warnings.warn(
                "The parameter 'ido' is depreciated and will be removed in PyPDF2 3.0.0.",
                DeprecationWarning,
            )
```

File: PyPDF2/_reader.py [hunk 2]
```diff
@@ -815,20 +815,20 @@ def threads(self) -> Optional[ArrayObject]:
...
 
     def _get_page_number_by_indirect(
-        self, indirect_ref: Union[None, int, NullObject, IndirectObject]
+        self, indirect_reference: Union[None, int, NullObject, IndirectObject]
     ) -> int:
         """Generate _page_id2num"""
         if self._page_id2num is None:
             self._page_id2num = {
-                x.indirect_ref.idnum: i for i, x in enumerate(self.pages)  # type: ignore
+                x.indirect_reference.idnum: i for i, x in enumerate(self.pages)  # type: ignore
```
Comment (MartinThoma) [inline PyPDF2/_reader.py:823]:
```suggestion
                x.indirect_ref.idnum: i for i, x in enumerate(self.pages)  # type: ignore
```

File: PyPDF2/_reader.py [hunk 3]
```diff
@@ -841,7 +841,7 @@ def get_page_number(self, page: PageObject) -> int:
             an instance of :class:`PageObject<PyPDF2._page.PageObject>`
         :return: the page number or -1 if page not found
         """
-        return self._get_page_number_by_indirect(page.indirect_ref)
+        return self._get_page_number_by_indirect(page.indirect_reference)
```
Comment (MartinThoma) [inline PyPDF2/_reader.py:844]:
```suggestion
        return self._get_page_number_by_indirect(page.indirect_ref)
```

File: PyPDF2/_reader.py [hunk 4]
```diff
@@ -904,10 +904,10 @@ def _build_destination(
                 if self.strict:
                     raise
                 # create a link to first Page
-                tmp = self.pages[0].indirect_ref
-                indirect_ref = NullObject() if tmp is None else tmp
+                tmp = self.pages[0].indirect_reference
```
Comment (MartinThoma) [inline PyPDF2/_reader.py:907]:
```suggestion
                tmp = self.pages[0].indirect_ref
```

File: tests/test_writer.py [hunk 5]
```diff
@@ -830,10 +830,10 @@ def test_startup_dest():
     pdf_file_writer.open_destination = pdf_file_writer.pages[9]
     # checked also using Acrobrat to verify the good page is opened
     op = pdf_file_writer._root_object["/OpenAction"]
-    assert op[0] == pdf_file_writer.pages[9].indirect_ref
+    assert op[0] == pdf_file_writer.pages[9].indirect_reference
```
Comment (MartinThoma) [inline tests/test_writer.py:833]:
```suggestion
    assert op[0] == pdf_file_writer.pages[9].indirect_ref
```

File: tests/test_writer.py [hunk 6]
```diff
@@ -830,10 +830,10 @@ def test_startup_dest():
...
     assert op[1] == "/Fit"
     op = pdf_file_writer.open_destination
-    assert op.raw_get("/Page") == pdf_file_writer.pages[9].indirect_ref
+    assert op.raw_get("/Page") == pdf_file_writer.pages[9].indirect_reference
```
Comment (MartinThoma) [inline tests/test_writer.py:836]:
```suggestion
    assert op.raw_get("/Page") == pdf_file_writer.pages[9].indirect_ref
```

**Review by MartinThoma — COMMENTED** [review]
Rename `Page.indirect_reference` temporarily back to `Page.indirect_ref` to fix the PR

File: PyPDF2/_writer.py [hunk 7]
```diff
@@ -456,7 +465,7 @@ def open_destination(self, dest: Union[None, str, Destination, PageObject]) -> N
         elif isinstance(dest, PageObject):
             self._root_object[NameObject("/OpenAction")] = Destination(
                 "Opening",
-                dest.indirect_ref if dest.indirect_ref is not None else NullObject(),
+                dest.indirect_reference if dest.indirect_reference is not None else NullObject(),
```
Comment (MartinThoma) [inline PyPDF2/_writer.py:468]:
```suggestion
                dest.indirect_ref if dest.indirect_ref is not None else NullObject(),
```

File: PyPDF2/_writer.py [hunk 8]
```diff
@@ -201,10 +201,19 @@ def _add_object(self, obj: PdfObject) -> IndirectObject:
...
+        if ido is not None:
+            if indirect_reference is not None:
+                raise ValueError("Please only set 'indirect_reference'. The 'ido' argument is deprecated.")
+            else:
+                indirect_reference = ido
+                warnings.warn(
+                    "The parameter 'ido' is depreciated and will be removed in PyPDF2 3.0.0.",
+                    DeprecationWarning,
+                )
+        if indirect_reference.pdf != self:
```
Comment (MartinThoma) [inline PyPDF2/_writer.py:214]:
```suggestion
        assert indirect_reference is not None  # the None value is only there to keep the deprecated name
        if indirect_reference.pdf != self:
```

**Pull request merged**